Sources/App/QRCodeScanner/Camera/BarcodeScannerCamera.swift

```diff
@@ -0,0 +1,335 @@
+                guard success else { return }
+                self.captureSession.startRunning()
+            }
+        }
+    }
+
+    func stop() {
+        guard isCaptureSessionConfigured else { return }
+
+        if captureSession.isRunning {
+            sessionQueue.async {
+                self.captureSession.stopRunning()
+            }
+        }
+    }
+
+    func toggleFlashlight() {
+        guard let captureDevice, captureDevice.hasTorch else { return }
+
+        do {
+            try captureDevice.lockForConfiguration()
+
+            if captureDevice.torchMode == .off {
+                try captureDevice.setTorchModeOn(level: AVCaptureDevice.maxAvailableTorchLevel)
+            } else {
+                captureDevice.torchMode = .off
+            }
+
+            captureDevice.unlockForConfiguration()
+        } catch {
+            Current.Log.info("Flashlight could not be used: \(error)")
+        }
+    }
+
+    private var deviceOrientation: UIDeviceOrientation {
+        var orientation = UIDevice.current.orientation
+        if orientation == UIDeviceOrientation.unknown {
+            orientation = UIScreen.main.orientation
+        }
+        return orientation
+    }
+
+    private func videoOrientationFor(_ deviceOrientation: UIDeviceOrientation) -> AVCaptureVideoOrientation? {
+        switch deviceOrientation {
+        case .portrait: return AVCaptureVideoOrientation.portrait
+        case .portraitUpsideDown: return AVCaptureVideoOrientation.portraitUpsideDown
+        case .landscapeLeft: return AVCaptureVideoOrientation.landscapeRight
+        case .landscapeRight: return AVCaptureVideoOrientation.landscapeLeft
+        default: return nil
+        }
+    }
+}
+
+extension BarcodeScannerCamera: AVCaptureVideoDataOutputSampleBufferDelegate {
+    func captureOutput(
+        _ output: AVCaptureOutput,
+        didOutput sampleBuffer: CMSampleBuffer,
+        from connection: AVCaptureConnection
+    ) {
+        guard let pixelBuffer = sampleBuffer.imageBuffer else { return }
+
+        if connection.isVideoOrientationSupported,
+           let videoOrientation = videoOrientationFor(deviceOrientation) {
+            connection.videoOrientation = videoOrientation
+        }
+
+        addToPreviewStream?(CIImage(cvPixelBuffer: pixelBuffer))
+    }
+}
+
+extension BarcodeScannerCamera: AVCaptureMetadataOutputObjectsDelegate {
+    func metadataOutput(
+        _ output: AVCaptureMetadataOutput,
+        didOutput metadataObjects: [AVMetadataObject],
+        from connection: AVCaptureConnection
+    ) {
+        if let metadataObject = metadataObjects.first {
+            let format = metadataObject.type.haString
+            guard let readableObject = metadataObject as? AVMetadataMachineReadableCodeObject else { return }
+            guard let stringValue = readableObject.stringValue else { return }
+            AudioServicesPlaySystemSound(SystemSoundID(kSystemSoundID_Vibrate))
```

Comment:
why this over haptic engine's success haptic?

Reply:
I was having an issue where the haptics were "glitching" while the view controller dismisses, difficult to explain in writting, but basically it was vibrating severald times while dismissing.
It endup being my test device, after rebooting it it's good now. I have updated the implementation